### PR TITLE
Philiphs Hue Improvments

### PR DESCRIPTION
**Pull request by brindosch**
**1.** Tell us something about your changes.
This is a patched, rewriten and improved version of the philiphs hue implementation
Changes:
- segfaults segfault segfaults. No segfaults! Solid against wrong configuration, bridge not reachable, unknown light ids and so on
- No thread freezers anymore which leads to unexpected behaviour like closed sockets or not responding sockets of the hyperion api
- check if bridge is reachable, user has access
- latchTime auto adjusted based on configured lightIDs
- switchOffOnBlack working again
- compatible with led device component on/off
- more logging what's actually going on
- lightID detection should be better, i hope that all bridges supports this (update to recent firmware to be save)
- 2 new lamp models added

@ntim 
xy == CiColor::BLACK -> rgb 0,0,0 is never CiColor 0,0,0, i got 0.164, 0.04, 0 i switched now to check just against bri, that seems to work well.

Testers wanted
@Gamer06 @b1rdhous3 @seblagaffe

**2.** If this changes affect the .conf file. Please provide the changed section

**3.** Reference an issue (optional)
https://github.com/hyperion-project/hyperion.ng/issues/466
https://github.com/hyperion-project/hyperion.ng/issues/429

Note: For further discussions use our forum: forum.hyperion-project.org

### CODE REVIEW

**Comment by brindosch** [comment]
I think there is a brightness calculation issues of blue (at least)
red (255,0,0) leads to brightness  72/254 (!)
green (0,255,0) leads to brightness 170/254
blue (0,0,255) lead to brightness 12/254 (!)

You can watch the brightness slider at the PhilipsHue App for better traceability. Don't forget to set the brightness compensation to 0% on the color page if you didn't used my webui philipshue assistant for configuration (available at led hardware -> philipshue)

So please report your results.

**Comment by redPanther** [comment]
Great! I wished I had a hue ....

File: libsrc/leddevice/dev_net/LedDevicePhilipsHue.cpp [hunk 1]
```diff
@@ -126,95 +122,134 @@ float CiColor::getDistanceBetweenTwoPoints(CiColor p1, CiColor p2)
...
+				Error(log, "Got invalid response from bridge");
+				return;
+			}
+			// check for authorization
+			if(doc.isArray())
+			{
+				Error(log, "Authorization failed, username invalid");
+				return;
+			}
+			else
```
Comment (redPanther) [inline libsrc/leddevice/dev_net/LedDevicePhilipsHue.cpp]: Code lile
If ()
{
 return
}
else
{
action
}

Can be written as:
If ()
{
Return
}

action

No need fot 'else' branch

Comment (brindosch) [reply libsrc/leddevice/dev_net/LedDevicePhilipsHue.cpp]: Of course, now where you name it

**Comment by brindosch** [comment]
@redPanther 
I added a new LedDevice::switchOn() method, as i want a new initialization if you enable the led device. This removes a bool workaround in my write() function
Hope this is ok

**Comment by redPanther** [comment]
Hmm I'm not sure this is the best solution. Perhaps it is better to define signals when enable state is changed. Then connect to the signals if needed

**Comment by brindosch** [comment]
You mean one signal (and pass the bool state)?
The smartest way to get a pointer to the base class is?
 - static_cast at the child device constructor: LedDevice* md = static_cast<LedDevice*>(this);
 - method at base class that returns a pointer to it (eg getInstance())

Right?

**Comment by redPanther** [comment]
no, I mean:
```
void LedDevice::setEnable(bool enable)
{
	// emit signal when state changed
	if (_enabled != enable)
	{
		emit enableStateChanged(enable);
	}

	// set black to leds when they should go off
	if ( _enabled && !enable)
	{
		switchOff();
	}
	_enabled = enable;
}
```

and add: `signal  enableStateChanged(bool);` in header.

in derived class you can connect to this signal with a slot. This is a bit cleaner and means more what you want. You can use this signal outside the leddevices too. E.g. when a leddevices deactivates itselfe due to an error (e.g. network unreachable ...)

**Comment by brindosch** [comment]
So you want to keep the switchOff function? The signal replaces the functionality!? The write(BLACK) isn't a big deal.

To be more clear, to connect to the signal i need a pointer of the base class(?)
This works, but is it the way to go? esp as child class..
```
	LedDevice* ld = static_cast<LedDevice*>(this);
	connect(ld, &LedDevice::ledDeviceStateChanged, this, &LedDevicePhilipsHue::stateChanged);
```

**Comment by redPanther** [comment]
`connect(this, SIGNAL(enableStateChanged(bool)), this,  SLOT(stateChanged));`
cast should not needed because we in a derivided class.

I want to keep switch off func. (or rename to setBlack ...).

**Comment by brindosch** [comment]
Hm, now i ask myself why i didn't tried a second this :) But makes sense. Will do so... :)

**Comment by brindosch** [comment]
Thank you for your feedback

**Pull request merged**